Remove OpenSSL dependencies (#12411)
gravitational/teleport: docs/pages/desktop-access/getting-started.mdx

  ![Deny Interactive Login](../../img/desktop-access/deny-interactive-login.png)
</Figure>

> ## Step 3/7: Configure a GPO to allow Teleport connections

Reviewer: ```suggestion
## Step 3/7. Configure a GPO to allow Teleport connections
```

For consistency with other steps

  ![Disable Require](../../img/desktop-access/disable.png)
</Figure>

> ### Configure certificate for RDP connections

Author: Yes, you're right. I've added a sentence to the admonition at the beginning of that chapter that states that it is only necessary for Windows Server 2012 R2. Do you think it is enough or should I add it more explicitly?

> ### Configure certificate for RDP connections

Reviewer 2: @ptgott can advise us on this

> ### Configure certificate for RDP connections

Reviewer 2: @LKozlowski it would also be good to confirm that if you make these changes and try to run with an old (openssl) version of Teleport that it still works.

> ### Configure certificate for RDP connections

Author: @zmb3 it works on both version with new certificates

<Admonition type="note" title="Secure Cipher Suites">
  Teleport's RDP client supports only secure algorithms
  for making TLS connections, so we have to configure a Windows server
  to support those cipher suites as well.
</Admonition>

In this step we'll create a new Certificate Authority's certificate template
with a certificate that uses elliptic curves cryptography and then configure

> our GPO to use the newly created template for Remote Desktop connections.

Author: Thanks for suggestion!

  ![Disable Require](../../img/desktop-access/disable.png)
</Figure>

### Configure certificate for RDP connections

<Admonition type="note" title="Secure Cipher Suites">
  Teleport's RDP client supports only secure algorithms
  for making TLS connections, so we have to configure a Windows server
  to support those cipher suites as well.

>   This step is only *neccessary* for Windows Server 2012 R2 as it does not support

Reviewer 3: ```suggestion
  This step is only *necessary* for Windows Server 2012 R2 as it does not support
```

  secure algorithms by default.
</Admonition>

> 

Reviewer 3: nit:

```suggestion
```

In this step we'll create a new certificate template that uses
elliptic curve cryptography, and then configure our GPO

> to use the newly created template to issue certificates used for Remote Desktop connections.

Reviewer 3: nit: Let's pick a line length and use it consistently. Whatever the rest of the file does.

> # Create a certificate template

Reviewer 3: Let's give a brief description of what we are trying to achieve here, before the steps. Ie: "we need to create a certificate template with parameters X and Y".

Knowing the intent can be helpful if the UI changes, or if someone is doing the setup in a slightly different way. It also highlights what is required from steps we just have to do to get there.

> 1. Begin by navigating to:
> ```
> Start > Control Panel > Administrative Tools > Certificate Authority
> ```
> to open the CA Microsoft Management Console (MMC) GUI.

Reviewer 3: Maybe:

```suggestion
1. Open the Microsoft Management Console (MMC):

    ```
    Start > Control Panel > Administrative Tools > Certificate Authority
    ```
```

so we avoid the line breaks, the way it is it'll probably mess with the list items.

Author: Yeah this looks way better, thanks!

2. Open your CA computer and right-click on `Certificate Templates`, then select `Manage`.

> 3. Find `Computer` template on the list, right-click on it, then select `Duplicate Template`.

Reviewer 3: Possibly?

```suggestion
3. Find the `Computer` template on the list, right-click on it, then select `Duplicate Template`.
```

> 4. In the `Compatibility` tab change `Certification Authority` to `Windows Server 2012 R2` and click `OK`.
> In the same tab change `Certificate recipient` to `Windows Server 2012 R2` and click `OK`.

Reviewer 3: Indent items inside the list to avoid breaking it. I would also space the contents, keeping the lines together breaks some markdown parsers.

```suggestion
4. In the `Compatibility` tab change `Certification Authority` to `Windows Server 2012 R2` and click `OK`.

    In the same tab change `Certificate recipient` to `Windows Server 2012 R2` and click `OK`.

```

5. Go to `General` tab and change `Template display name` to `RemoteDesktopAccess`. Make sure `Template name` is also `RemoteDesktopAccess`.

> 6. In the `Cryptography` tab change `Provider Category` to `Key Storage Provider`, then `Algorithm name` to `ECDH_P384`. Also, change `Request hash` to `SHA384`.

Reviewer 3: Does it have to be P384/SHA384 or did you pick those because they are the best available? Is there a lower bound?

No need to change the text, I'm just curious. Although, if there is a lower bound, let's mention it in the description that precedes this (ie, "certificates must be P384 or better").

Author: Yeah, it have to be this one. I picked it as it is enabled be default by Windows Server 2012 R2 and is also supported by rustls library. The other one that would also work is P521/SHA384 but it would require user to enable it manually (few extra steps that doesn't provide much of difference in our case).

7. Next, in the `Extensions` tab select `Application Polices` and click `Edit` button.
8. Remove all entries from the list.

> 9. Click `Add…` button, then `New…`. Set the value of `Name` to `Remote Desktop Authentication`. Change the object identifier to `1.3.6.1.4.1.311.54.1.2.` and click `OK`.

Reviewer 3: > (...) Change the object identifier to `1.3.6.1.4.1.311.54.1.2.`

What does `1.3.6.1.4.1.311.54.1.2.` mean? Could you add its "human" name in parenthesis, so people understand better what they are doing?

Author: This number is object identifier (OID) for Remote Desktop Access enhanced key usage extension, but this point will be removed completely. It turns out that we actually don't need to set it to make RDP working. The suggestion was good anyway. The human readable name should've been added if this point had to remain.

10. From the application policies list select `Remote Desktop Authentication` and click `OK`.
11. Go to `Security` tab, select `Domain Controllers` and give the group `Read` and `Enroll` permissions.
12. Finally, create template by clicking `OK`

> 13. Go back to the Certificate Authority window and right-click on `Certificate Templates`. Then:
> ```
> New > Certificate Template to Issue
> ```
> Select `RemoteDesktopAccess` and click `OK`.

Reviewer 3: Space and indent item contents.

```suggestion
13. Go back to the Certificate Authority window and right-click on `Certificate Templates`. Then:

    ```
    New > Certificate Template to Issue
    ```

    Select `RemoteDesktopAccess` and click `OK`.
```

> # Update GPO to use a new certificate template

Reviewer 3: What does GPO mean? Could you define the term the first time it is mentioned? (Step 2/6 I think.)

Ie, "blabla GPO (Gnome Party Onslaught)" or whatever it is.  :)

Author: GPO means Group Policy Object. We mention about it for the first time in step 2/6. It is also described at this step and then in the following steps we only use the abbreviation. Do you think the abbreviation should be explained here as well?

Reviewer 3: I missed the explanation in step 2. It would mention it there in the first occurrence instead of down the line, but I think the text is good here.

In the group policy editor for `Teleport Access Policy`, select:

```
Computer Configuration > Policies > Administrative Templates > Windows Components > Remote Desktop Services > Remote Desktop Session Host > Security
```

Right-click on `Server authentication certificate template`, `Edit`, then select `Enabled` and fill `Certificate Template Name` with `RemoteDesktopAccess`.

<Figure align="left" bordered caption="RDP Certificate Template">
  ![RDP Certificate Template](../../img/desktop-access/rdp-certificate-template.png)
</Figure>

> 

Reviewer 3: @zmb3 do you have a Windows in hand? Could you give the guide a try?

(Or anyone else that happens to be in the review, really.)

  ![Disable Require](../../img/desktop-access/disable.png)
</Figure>

> ### Configure certificate for RDP connections

Reviewer: It looks like this is the final subsection of "Step 3/6: Configure a GPO to allow Teleport connections". I'm not sure that's the intention, though, since the second paragraph begins with "In this step we'll...". 

If this is the beginning of a new step, we'll need to promote this section heading to an H2 (using `##` instead of `###`), add "Step n/d:" to the title, and renumber the rest of the H2 section headings (e.g., this would be "Step 4/7").

Author: I was thinking about whether I should add it as a separate step or add it to the existing one as it's related to it. After I gave it some thought and re-read it again, I agree that it make more sense to make it a separate optional step.

<Admonition type="note" title="Secure Cipher Suites">
  Teleport's RDP client supports only secure algorithms

>   for making TLS connections, so we have to configure a Windows server

Reviewer: Is "a Windows server" one that we've referred to earlier in this guide? Is it the one where we've been configuring our service account and GPOs? We might want to specify which host commands in this section will be run on.

Author: Yes this is the same Windows server. I'll add a clarification.

  to support those cipher suites as well.
  This step is only *necessary* for Windows Server 2012 R2 as it does not support

>   secure algorithms by default.

Reviewer: If we're advising some users to skip this step, I think we should add a link to the next step here.

Author: Great suggestion! It'll make it a bit easier for users to follow the steps in our guide.

Author: Is there a way to know what will be the link to the header? I see that the links are auto-generated based on the text, so I went to the current live docs and took it from there. I was wondering if there is a better way to do it?

Reviewer 3: That's what I do too. If you wanted to be explicit you could stick in some plain HTML  (`<h2 id="myanchor">Text</h2>`), but I think the preference is for markdown.

Reviewer 3: Btw, you can render locally using https://github.com/gravitational/next - either checkout or create your branch in the submodule within the sources.

</Admonition>

In this step we'll create a new certificate template that uses elliptic curve cryptography, and then configure our GPO
to use the newly created template to issue certificates used for Remote Desktop connections.

> ## Create a certificate template

Reviewer: ```suggestion
### Create a certificate template
```
Guessing this should be an H3 if the previous section was meant to be an H2.

> In this step we create a certificate template that uses elliptic curve P-384 and uses SHA384 as the signature algorithm.

Reviewer: ```suggestion
In this section, we will create a certificate template that uses elliptic curve P-384 and uses SHA384 as the signature algorithm.
```
To distinguish the beginning of this paragraph from the beginning of the last paragraph. The reader may be wondering, "when does the step begin?"

1. Open the Microsoft Management Console (MMC)

```text
Start > Control Panel > Administrative Tools > Certificate Authority
```

2. Open your CA computer and right-click on `Certificate Templates`, then select `Manage`.
3. Find the `Computer` template on the list, right-click on it, then select `Duplicate Template`.
4. In the `Compatibility` tab change `Certification Authority` to `Windows Server 2012 R2` and click `OK`.
5. In the same tab change `Certificate recipient` to `Windows Server 2012 R2` and click `OK`.

> 6. Go to `General` tab and change `Template display name` to `RemoteDesktopAccess`. Make sure `Template name` is also `RemoteDesktopAccess`.

Reviewer: ```suggestion
6. Go to the `General` tab and change `Template display name` to `RemoteDesktopAccess`. Make sure `Template name` is also `RemoteDesktopAccess`.
```

7. In the `Cryptography` tab change `Provider Category` to `Key Storage Provider`, then `Algorithm name` to `ECDH_P384`. Also, change `Request hash` to `SHA384`.
8. Next, in the `Extensions` tab select `Application Polices` and click `Edit` button.
9. Remove all entries from the list.
10. Go to `Security` tab, select `Domain Controllers` and give the group `Read` and `Enroll` permissions.

> 11. Finally, create template by clicking `OK`

Reviewer: ```suggestion
11. Finally, create a template by clicking `OK`.
```

12. Go back to the Certificate Authority window and right-click on `Certificate Templates`. Then:

```text
New > Certificate Template to Issue
```
Select `RemoteDesktopAccess` and click `OK`.

> ## Update GPO to use a new certificate template

Reviewer: ```suggestion
### Update GPO to use a new certificate template
```

  ![Disable Require](../../img/desktop-access/disable.png)
</Figure>

### Configure certificate for RDP connections

<Admonition type="note" title="Secure Cipher Suites">
  Teleport's RDP client supports only secure algorithms
  for making TLS connections, so we have to configure a Windows server
  to support those cipher suites as well.
  This step is only *necessary* for Windows Server 2012 R2 as it does not support
  secure algorithms by default.
</Admonition>

In this step we'll create a new certificate template that uses elliptic curve cryptography, and then configure our GPO
to use the newly created template to issue certificates used for Remote Desktop connections.

# Create a certificate template

> In this step we will create a certificate template that is using elliptic curve P-384 and is using SHA384 as an signature algorithim.

Author: thanks for suggestions!

1. Open the Microsoft Management Console (MMC)

```
Start > Control Panel > Administrative Tools > Certificate Authority

> ```

Author: Sure, I've updated it 4-space indents.

gpupdate.exe /force
```

> ## Step 4/7: Configure certificate for RDP connections

Reviewer: ```suggestion
## Step 4/7. Configure a certificate for RDP connections
```

to use the newly created template to issue certificates used for Remote Desktop connections.

### Create a certificate template

In this section, we will create a certificate template that uses elliptic curve P-384 and uses SHA384 as the signature algorithm.

1. Open the Microsoft Management Console (MMC)

```text
Start > Control Panel > Administrative Tools > Certificate Authority
```

2. Open your CA computer and right-click on `Certificate Templates`, then select `Manage`.
3. Find the `Computer` template on the list, right-click on it, then select `Duplicate Template`.
4. In the `Compatibility` tab change `Certification Authority` to `Windows Server 2012 R2` and click `OK`.
5. In the same tab change `Certificate recipient` to `Windows Server 2012 R2` and click `OK`.
6. Go to the `General` tab and change `Template display name` to `RemoteDesktopAccess`. Make sure `Template name` is also `RemoteDesktopAccess`.
7. In the `Cryptography` tab change `Provider Category` to `Key Storage Provider`, then `Algorithm name` to `ECDH_P384`. Also, change `Request hash` to `SHA384`.

> 8. Next, in the `Extensions` tab select `Application Polices` and click `Edit` button.

Reviewer: ```suggestion
8. Next, in the `Extensions` tab select `Application Polices` and click the `Edit` button.
```

9. Remove all entries from the list.

> 10. Go to `Security` tab, select `Domain Controllers` and give the group `Read` and `Enroll` permissions.

Reviewer: ```suggestion
10. Go to the `Security` tab, select `Domain Controllers` and give the group `Read` and `Enroll` permissions.
```
